nemoguardrails/library/autoguard/actions.py

```diff
@@ -0,0 +1,496 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import os
+from typing import Optional
+
+import aiohttp
+
+from nemoguardrails.actions import action
+
+log = logging.getLogger(__name__)
+
+
+@action(name="call autoguard gender bias api", is_system_action=True)
+async def call_autoguard_gender_bias_api(context: Optional[dict] = None):
+    api_key = os.environ.get("AUTOGUARD_API_KEY")
+
+    if api_key is None:
+        raise ValueError("AUTOGUARD_API_KEY environment variable not set.")
+
+    user_message = context.get("user_message")
+    bot_message = context.get("bot_message")
+    if bot_message:
+        prompt = bot_message
+    else:
+        prompt = user_message
+
+    print(prompt)
+
+    model = context.get("model")
+    url = "http://35.225.99.81:8888/"
+
+    if model:
+        url = url + "query"
+    else:
+        url = url + "guardrail"
+
+    headers = {"x-api-key": api_key}
+
+    data = {
+        "config": {
+            "tonal_detection": {"mode": "OFF"},
+            "pii_fast": {
+                "mode": "OFF",
+                "enabled_types": [],
+                "mask": False,
+                "coreference": False,
+            },
+            "factcheck": {"mode": "OFF"},
+            "confidential_detection": {"mode": "OFF"},
+            "jailbreak_detection": {"mode": "OFF"},
+            "text_toxicity_extraction": {"mode": "OFF"},
+            "harm_detection": {"mode": "OFF"},
+            "racial_bias_detection": {"mode": "OFF"},
+            "intellectual_property": {"mode": "OFF"},
+            "gender_bias_detection": {"mode": "DETECT"},
+        },
+        "prompt": prompt,
+    }
+    if model:
+        data["model"] = model
+    async with aiohttp.ClientSession() as session:
+        async with session.post(
+            url=url,
+            headers=headers,
+            json=data,
+        ) as response:
+            if response.status != 200:
+                raise ValueError(
+                    f"AutoGuard call failed with status code {response.status}.\n"
+                    f"Details: {await response.text()}"
+                )
+            data = await response.text()
+            response_json = []
+            for line in data.split("\n"):
```

Comment:
since we stream out the response as lines, best to receive as lines instead of parsing it later

https://stackoverflow.com/questions/59681726/how-to-read-lines-of-a-streaming-api-with-aiohttp-in-python

Reply:
the streaming only applies to /guardrail, /query has a different response and should be treated differently overall

Reply:
resolved in previous commits